cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -670,9 +671,11 @@ Status NumPyConverter::Visit(const StringType& type) {
         return Status::Invalid("Encountered non-UTF8 binary value: ",
                                HexEncode(data, itemsize_));
       }
-    } else {
+    } else if (is_unicode_type) {
```

Comment:
Instead of checking this condition inside the append loop, you can raise the error outside (by checking `!is_binary_type && !is_unicode_type`).